src/Build/Graph/GraphBuilder.cs

```diff
@@ -0,0 +1,397 @@
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Build.BackEnd;
+using Microsoft.Build.Collections;
+using Microsoft.Build.Evaluation;
+using Microsoft.Build.Execution;
+using Microsoft.Build.Shared;
+
+namespace Microsoft.Build.Experimental.Graph
+{
+    internal class GraphBuilder
+    {
+        private ProjectInterpretation ProjectInterpretation { get; }
+
+        public IReadOnlyCollection<ProjectGraphNode> ProjectNodes { get; private set; }
+
+        public IReadOnlyCollection<ProjectGraphNode> RootNodes { get; private set; }
+
+        public IReadOnlyCollection<ProjectGraphNode> EntryPointNodes { get; private set; }
+
+        public GraphEdges Edges { get; private set; }
+
+        public GraphBuilder(ProjectInterpretation projectInterpretation)
+        {
+            ProjectInterpretation = projectInterpretation;
+        }
+
+        public void BuildGraph(
+            IEnumerable<ProjectGraphEntryPoint> entryPoints,
+            ProjectCollection projectCollection,
+            ProjectGraph.ProjectInstanceFactoryFunc projectInstanceFactory
+            )
+        {
+            var entryPointConfigurationMetadata = AddGraphBuildPropertyToEntryPoints(entryPoints);
+
+            var allParsedProjects = new ConcurrentDictionary<ConfigurationMetadata, ProjectGraphNode>();
+            var tasksInProgress = new ConcurrentDictionary<ConfigurationMetadata, object>();
+            var projectsToEvaluate = new ConcurrentQueue<ConfigurationMetadata>();
+
+            foreach (var entryPoint in entryPointConfigurationMetadata)
+            {
+                projectsToEvaluate.Enqueue(entryPoint);
+            }
+
+            if (FindGraphNodes(
+                projectsToEvaluate,
+                projectCollection,
+                tasksInProgress,
+                projectInstanceFactory,
+                ProjectInterpretation,
+                allParsedProjects,
+                out var exceptions))
+            {
+                Edges = new GraphEdges();
+
+                CreateEdgesAndDetectCycles(entryPointConfigurationMetadata, ProjectInterpretation, allParsedProjects, Edges);
+
+                ProjectInterpretation.PostProcess(allParsedProjects, this);
+
+                EntryPointNodes = entryPointConfigurationMetadata.Select(e => allParsedProjects[e]).ToList();
+                RootNodes = GetGraphRoots(EntryPointNodes);
+                ProjectNodes = allParsedProjects.Values.ToList();
+            }
+            else
+            {
+                throw new AggregateException(exceptions);
+            }
+
+            IReadOnlyCollection<ProjectGraphNode> GetGraphRoots(IReadOnlyCollection<ProjectGraphNode> entryPointNodes)
+            {
+                var graphRoots = new List<ProjectGraphNode>(entryPointNodes.Count);
```

Comment:
Ok to waste memory here if not all entryPoints end up being root?

Reply:
Trimmed the excess memory.